File: testing-kit/helpers/factories.ts
// Generic test data factories — callable as makeX(overrides).
// Use freely; copy and adapt to your project's domain types.

let idCounter = 0;
export function nextId(prefix = 'id'): string {
  return `${prefix}-${++idCounter}`;
}

export function resetIdCounter(): void { idCounter = 0; }

/** Build an ISO timestamp at an offset (ms) from a base. */
export function isoOffset(baseMs: number, offsetMs: number): string {
  return new Date(baseMs + offsetMs).toISOString();
}

/** Stable "now" for tests — a known UTC date you can pin via vi.setSystemTime. */
export const FIXED_NOW = new Date('2026-04-25T12:00:00.000Z').getTime();

/** Build N items with a per-index transform. */
export function buildMany<T>(n: number, build: (i: number) => T): T[] {
  return Array.from({ length: n }, (_, i) => build(i));
}

/** Pick one item from an array based on an index (mod). Useful for cycling priorities/categories. */
export function cycle<T>(arr: readonly T[], i: number): T {
  return arr[i % arr.length];
}

/** Deep merge for plain objects/arrays — used inside factories for partial overrides. */
export function deepMerge<T>(base: T, over: any): T {
  if (over === undefined || over === null) return base;
  if (Array.isArray(base) || Array.isArray(over) || typeof base !== 'object') return over as T;
  const out: any = { ...base };
  for (const k of Object.keys(over)) {
    out[k] = deepMerge((base as any)[k], over[k]);
  }
  return out as T;
}

/** Snapshot a Zustand-ish store length object — useful in "tour does not pollute store" tests. */
export function snapshotLengths(state: Record<string, unknown>): Record<string, number> {
  const out: Record<string, number> = {};
  for (const k of Object.keys(state)) {
    const v = (state as any)[k];
    if (Array.isArray(v)) out[k] = v.length;
  }
  return out;
}

/** Wait for a predicate without polling-busywait. */
export async function waitFor(predicate: () => boolean, timeoutMs = 1000, intervalMs = 10): Promise<void> {
  const start = Date.now();
  while (!predicate()) {
    if (Date.now() - start > timeoutMs) throw new Error('waitFor timed out');
    await new Promise((r) => setTimeout(r, intervalMs));
  }
}

File: testing-kit/helpers/factories.py
"""Generic Python test factories — small helpers for building fixtures fast."""
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from itertools import count
from typing import Callable, Iterable, TypeVar

T = TypeVar("T")

_id_counter = count(1)
FIXED_NOW = datetime(2026, 4, 25, 12, 0, 0, tzinfo=timezone.utc)


def next_id(prefix: str = "id") -> str:
    return f"{prefix}-{next(_id_counter)}"


def reset_id_counter() -> None:
    global _id_counter
    _id_counter = count(1)


def iso_offset(base: datetime, offset_seconds: float) -> str:
    return (base + timedelta(seconds=offset_seconds)).isoformat()


def build_many(n: int, build: Callable[[int], T]) -> list[T]:
    return [build(i) for i in range(n)]


def cycle(items: Iterable[T], i: int) -> T:
    arr = list(items)
    return arr[i % len(arr)]


def deep_merge(base, over):
    """Recursive merge of two plain dicts. Right-hand wins on scalars/arrays."""
    if not isinstance(base, dict) or not isinstance(over, dict):
        return over
    out = dict(base)
    for k, v in over.items():
        out[k] = deep_merge(base.get(k), v)
    return out


def snapshot_lengths(state: dict) -> dict[str, int]:
    return {k: len(v) for k, v in state.items() if isinstance(v, (list, tuple, set, dict))}
